tools/taskcluster-worker-runner/protocol/protocol.go

```diff
@@ -12,6 +15,9 @@ type Protocol struct {
 	// to avoid finding the empty set at startup)
 	Capabilities *Capabilities
 
+	// Capabilities of the remote peer
+	RemoteCapabilities *Capabilities
```

Comment:
I like splitting this into "our" capabilities and the remote system's capabilities.  However, this loses the effect of intersecting the two sets of capabilities to find out what both ends have in common.  So for example a worker using this protocol would unconditionally list `"shutdown"` in its capabilities, and call `proto.Capable("shutdown")` to find out whether it's supported, which would return true.  Even if the worker-runner it was talking to did not support shutdown.

Maybe `proto.Capable` should check both sets of capabilities and && the results?

Reply:
Makes sense.

Reply:
Hrm, I gave a further thought on this. Not necessarily an unsupported capability of the worker will prevent worker-runner from making it available. `shutdown` is such an example. I believe this level of graining check should be a layer upper.